V2 - many improvements, new release cycle

tomzo opened the pull request:
The goal of this PR is to make the plugin capable of replacing original GoCD artifact store.

The new readme explains new features and behavior in detail.

### Summary of changes

Features and improvements:
 - Added numerous tests at unit and integration level
 - Implemented pattern maching and directory uploads
 - Implemented fetch of multiple files or directories
 - Implemented environment variable expansion in the destination prefix of the S3 bucket

Build system and release cycle:
 - Updated build system to use gradle 5.x
 - Added builds in docker using dojo.
 - Automated testing and releases
 - Added integration tests with AWS S3 bucket, publish and fetch cycle

## Review Notes

@Diogomrol all of this is in line with what we discussed earlier. It should be easy to review commit-by-commit because each does only 1 thing.

These changes are breaking compatibility with previous version, because the JSON object stored in GoCD now has 3 fields, which are required to fetch properly.

Once this is merged, my GoCD server can make a new release after manual approval. Currently I have released this on [my fork](https://github.com/tomzo/gocd-s3-artifact-plugin/releases) as a preview.

I'll be trying out the plugin starting today.

Let me know what you think. If you think something is not clear/right I can clarify or fix it.

tomzo committed:
update build system

- new gradle
- use gocd plugin helpers

tomzo committed:
dev env setup with terraform, real s3 bucket, gocd server and agent

tomzo committed:
fix publish unit test to actually verify the upload

tomzo committed:
remove content type metadata from uploaded objects

We have no certainty about the content of any artifact

tomzo committed:
added destination to artifact metadata

The value is the same as in artifact plan in the job, it does not
include the bucket name, which is part of store config.

tomzo committed:
refactoring and fetch single file unit test

tomzo committed:
implemented robust upload and fetch for multiple files #2 #10

tomzo committed:
implemented variable expansion for destination field

tomzo committed:
setup integration tests and ci build

tomzo committed:
ops security: don't print AWS secret key to the console output

tomzo committed:
docs on AWS policy

tomzo committed:
make travis build consistent with GoCD - use dojo

tomzo committed:
update gradle to 5.4.1

- Updated gradlew setup to use 5.4.1
- Updated dojo image to 1.1.0 which has gradle 5.4.1
- No need to use gradlew inside docker image, because it is already
bundled

tomzo committed:
setup release cycle on github

tomzo committed:
changelog for 2.0.0

tomzo committed:
fixup release task

Destination file should not include directory path

tomzo committed:
switch github release destination

tomzo committed:
added travis icon to the readme

Diogomrol commented on README.md:
```diff
@@ -120,6 +120,17 @@ Then there are no required options.
+
+One of the most important uses of this feature is when destination directory is specified by `${GO_PIPELINE_NAME}/${GO_PIPELINE_COUNTER}/${GO_STAGE_NAME}/${GO_STAGE_COUNTER}/${GO_JOB_NAME}`.
+Such destination is unique for each build. It is the same tree structure which original GoCD artifact store uses.
+This plugin provides a convenience `${GO_ARTIFACT_LOCATOR}` which evaluates to above expression.
```
Suggestion here: the behavior for GoCD is creating this destination directory without the user specifying it.

In order to keep some consistency of behavior, shall we do the same here? Although if we do then the user has less freedom on this :) Would just like to hear your thoughts @tomzo 👍

tomzo replied:
Yeah, I thought about it. I don't feel strongly way or another.
Now the plugin doesn't really behave exactly like GoCD did. Not just in this case.
 
But this is more powerful too. It can help implement other routines which previously were hard to do and required custom scripts.

I'd start with the way it is. With most freedom for the user. Maybe later we could add an option in artifact store configuration which would pre-append this to all destinations?

Diogomrol replied:
Yup, seems fair enough to me. Just wanted to understand your thoughts about this :) All seems great to me

Diogomrol reviewed (commented):
Added a small comment about the destination directory

Diogomrol commented:
@tomzo approved and everything merged! Will you take care of the release as we spoke?

tomzo commented:
Great! Thanks. 
I changed gocd setup and we should see first release in a few minutes.

tomzo commented:
The release is out.
There is just one more step which is not working...
I have implemented an automated bump, which would now push a changelog update to `2.0.1` and updated `build.gradle` with new version too.
That step is executed from [ai-go](https://github.com/ai-go) bot's account. Currently it gets permission denied when pushing the commit on master.
Do you mind adding [ai-go](https://github.com/ai-go) as collaborator too?

Diogomrol commented:
@tomzo done

tomzo commented:
Now push fails with:
```
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Required status check "Travis CI - Pull Request" is expected. At least 1 approving review is required by reviewers with write access.
To git@github.com:Diogomrol/gocd-s3-artifact-plugin.git
 ! [remote rejected] master -> master (protected branch hook declined)
error: failed to push some refs to 'git@github.com:Diogomrol/gocd-s3-artifact-plugin.git'
```
I think this check is unnecessary.
We can agree to not merge changes and PRs unless reviewed. I don't think this level of formality is helping.

Diogomrol commented:
@tomzo what is the issue? It has to do with the required approval?

tomzo commented:
I think you configured a required review somewhere, so that pushing directly to master is not possible without making a PR and accepting a review first.

Diogomrol commented:
@tomzo can you try again?

tomzo commented:
It's still failining
```
$ git push diogo master
Counting objects: 4, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 388 bytes | 0 bytes/s, done.
Total 4 (delta 3), reused 0 (delta 0)
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Required status check "Travis CI - Pull Request" is expected.
To git@github.com:Diogomrol/gocd-s3-artifact-plugin.git
 ! [remote rejected] master -> master (protected branch hook declined)
error: failed to push some refs to 'git@github.com:Diogomrol/gocd-s3-artifact-plugin.git'
```

Diogomrol commented:
Sorry, now it should work
